### DISCUSSION TITLE
From CAD (STL/STEP) through voxel to pore network - how to make it?

### DISCUSSION THREAD

**Discussion by pashaLyb** [PMEAL/OpenPNM#1888] [Q&A] [answered]
I'm new to OpenPNM and I would like to test it for PTL/GDL (porous transport layer/gas diffusion layer) characterization. Usually one will use µCT scan to build the p-network. I'm wondering if it is method to import a PTL structure defined by CAD geometry or STL/STEP file? This is interesting for a specific kind of the PTL/GDL that have a relatively coarse deterministic structure like metal net and could be easily defined/modeled with CAD and exported as STL or STEP file.

I copy here the answer by @jgostick  
> We currently do not have that, only the other way around (export to salome, comsol, stl, etc). There are two things you're actually asking about here though. One is to generate a voxel image out of a CAD file so that a network can be extracted. Two is importing a network that is defined by sticks and balls in CAD. Both are feasible but I've never 'looked' inside a cad file to see how the info is stored. Would you be able to share one? Also, feel free to start a fresh 'discussion' for this subject so we can have the thread in one place.

I'm more interested in first thing/topic you mentioned - "generate a voxel image out of a CAD file so that a network can be extracted"
Here are couple of examples I found in WEB

- [from CAD with FreeCAD](https://forum.freecadweb.org/viewtopic.php?t=30172)
- [Open Cascade](https://dev.opencascade.org/doc/occt-6.8.0/overview/html/occt_user_guides__voxels_wp.html)
- [Online Voxelizer](https://drububu.com/miscellaneous/voxelizer/?out=obj)

With the Online Voxelizer I converted a stl file I found im web to this voxel-model.

![image](https://user-images.githubusercontent.com/39120395/111800666-7904dd80-88cc-11eb-96d3-1ea39cbe11ff.png)

[something-inside20210314-23024-1w6tujn.zip](https://github.com/PMEAL/OpenPNM/files/6172123/something-inside20210314-23024-1w6tujn.zip)

My goal is to take STL or STEP --> transfer to any voxel format OpenPNM supports --> conduct simulations

Thanks

**Comment by pashaLyb** [comment]
maybe it's possible to utilizes this package? - [p-hofmann/PyVoxelizer](https://github.com/p-hofmann/PyVoxelizer)

**Reply by jgostick** [reply]
PyVoxelizer and FreeCAD both seem promising as they have python bindings (or are pure python).  I will play with them and report back.

**Reply by pashaLyb** [reply]
one more I found
[tasotasoso/STLSimpleVoxelizer](https://github.com/tasotasoso/STLSimpleVoxelizer)

**Reply by jgostick** [reply]
Yes, I'm finding quite a few as well. I'm trying to identify a stable and mature one.

**Reply by jgostick** [reply]
The [trimesh](https://github.com/mikedh/trimesh) library seems to be the best available.  I have confirmed that it has the ability to take an stl and convert to voxels, but the voxel conversion seems to be pretty slow.

**Comment by jgostick** [comment, marked as answer]
OK, here is a script that works.  I had to play a bit because the ``m.voxelized`` method of the ``mesh`` objects in ``trimesh`` only do the 'surface' mesh not the total volume.

``` python
import porespy as ps
from skimage.measure import marching_cubes
import scipy.ndimage as spim
import trimesh
import numpy as np


# Generate a basic image for testing
im1 = np.zeros(shape=[60, 60, 60], dtype=bool)
im1[tuple((slice(20, 40, None), slice(20, 40, None), slice(20, 40, None)))] = 1
im1[tuple((slice(30, 50, None), slice(30, 50, None), slice(30, 50, None)))] = 1

# Obtain a mesh from the boolean image.  
im1 = np.pad(im1, 3, mode='constant', constant_values=0)
im2 = spim.convolve(im1*1.0, weights=ps.tools.ps_round(1, im1.ndim, smooth=False))
im2 = im2/im2.max()
verts, faces, normals, values = marching_cubes(im2, level=0.5,
                                               gradient_direction='ascent')
m = trimesh.Trimesh(vertices=verts, faces=faces, vertex_normals=normals)

# If you have an stl file already, you could use:
# m = trimesh.load('filename.stl', file_type='stl')

# Finally, use the contains method of the mesh to see which points land within it
s = 1  # Increase this spacing to reduce image resolution
args = [range(0, im1.shape[i], s) for i in range(im1.ndim)]
pts = np.meshgrid(*args)
coords = np.vstack([pts[i].flatten() for i in range(im1.ndim)]).T
b = m.contains(coords)  # This step is not fast!!
b = b.reshape((np.array(im1.shape)/s).astype(int))
```
